### PR TITLE
ARROW-189: Build 3rd party with ExternalProject.

### PR DESCRIPTION
**Pull request by leifwalsh**
When third party env vars *_HOME are not present, use cmake's
ExternalProject to fetch and build them.  When those vars are present,
we just use them.

### CODE REVIEW

File: cpp/src/arrow/ipc/CMakeLists.txt [hunk 1]
```diff
@@ -84,13 +87,23 @@ foreach(FIL ${FBS_SRC})
...
+    VERBATIM
+  )
+else()
+  add_custom_command(
+    OUTPUT ${FBS_OUTPUT_FILES}
+    COMMAND ${FLATBUFFERS_COMPILER} -c -o ${OUTPUT_DIR} ${ABS_FBS_SRC}
+    DEPENDS ${ABS_FBS_SRC}
+    COMMENT "Running flatc compiler on ${ABS_FBS_SRC}"
+    VERBATIM
+  )
```
Comment (wesm) [inline cpp/src/arrow/ipc/CMakeLists.txt]: Any issues with setting the dependencies in a cmake variable and having one `add_custom_command` call?

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -476,18 +470,78 @@ include_directories(SYSTEM ${Boost_INCLUDE_DIR})
...
+    set(GBENCHMARK_STATIC_LIB "${GBENCHMARK_PREFIX}/lib/libbenchmark.a")
+    add_library(benchmark STATIC IMPORTED)
+    set_target_properties(benchmark PROPERTIES
+      IMPORTED_LOCATION ${GBENCHMARK_STATIC_LIB})
+    add_dependencies(benchmark gbenchmark_ep)
+    set(GBENCHMARK_VENDORED 1)
+  else()
+    find_package(GBenchmark REQUIRED)
+    ADD_THIRDPARTY_LIB(benchmark
+      STATIC_LIB ${GBENCHMARK_STATIC_LIB})
```
Comment (wesm) [inline cpp/CMakeLists.txt]: Possible to use `ADD_THIRDPARTY_LIB` in the ep portion?

**Review by wesm — COMMENTED** [review]
Can we remove the download_thirdparty.sh and build_thirdparty.sh scripts now? in theory then https://github.com/apache/arrow/blob/master/cpp/setup_build_env.sh may not be necessary in Travis CI

**Review by xhochy — CHANGES_REQUESTED** [review]
LGTM apart from @wesm comments. Once they're fixed I'm really pleased to see this merged.

File: cpp/src/arrow/ipc/CMakeLists.txt (hunk 1 above)
Comment (leifwalsh) [inline cpp/src/arrow/ipc/CMakeLists.txt]: sure

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (leifwalsh) [inline cpp/CMakeLists.txt]: yep, something else was broken when I switched away from it, it should work

**Comment by leifwalsh** [comment]
Ok, I think I've got travis as happy as it is on master.  Anything else?

**Comment by wesm** [comment]
Looks good. +1

**Pull request closed without merging**